### ISSUE TITLE
Including $libdir/pg_strom in the shared_preload_libraries throws error

### ISSUE THREAD

**Issue by pmdy** [heterodb/pg-strom#478]
Changed **/var/lib/pgsql/10/data/postgresql.conf** to accomodate `pg_strom` library. I only tuned for this parameter.

```conf
shared_preload_libraries = '$libdir/pg_strom'
```
Now, when I try to start the PostgreSQL server, I'm getting this error.

```bash
[root@3cac43ed98ff data]# systemctl start postgresql-10
Job for postgresql-10.service failed because the control process exited with error code. See "systemctl status postgresql-10.service" and "journalctl -xe" for details.
```

If I comment `shared_preload_libraries = '$libdir/pg_strom'`, PostgreSQL works again after restarting.

Dockerfile 
```Dockerfile
FROM nvidia/cuda:11.1-devel-centos7

ENV container docker

RUN yum -y install git wget unzip vim

RUN yum -y install https://dl.fedoraproject.org/pub/epel/7/x86_64/Packages/e/epel-release-7-13.noarch.rpm
RUN yum -y install https://heterodb.github.io/swdc/yum/rhel8-noarch/heterodb-swdc-1.1-1.el8.noarch.rpm

RUN yum -y install https://download.postgresql.org/pub/repos/yum/reporpms/EL-7-x86_64/pgdg-redhat-repo-latest.noarch.rpm
RUN yum install -y postgresql10-server postgresql10-devel postgresql10-contrib

RUN yum -y install postgresql-alternatives
RUN yum -y install pg_strom-PG10; exit 0

RUN wget https://github.com/heterodb/pg-strom/archive/v2.3.zip
RUN unzip v2.3.zip

WORKDIR /pg-strom-2.3
RUN make PG_CONFIG=/usr/pgsql-10/bin/pg_config
RUN make install PG_CONFIG=/usr/pgsql-10/bin/pg_config

RUN su postgres
RUN initdb -D /var/lib/pgdata/ 

# I manually do the remaining steps after above runs.
```

**Comment by kaigai** [comment]
Could you share which error happen on your environment? Typicaly, result of `journalctl -xe` or other logs.

**Comment by pmdy** [comment]
It looks like I have to install `nvme_strom` driver. Is it not optional?

```bash
[root@3cac43ed98ff data]# journalctl -xe
Nov 25 06:16:08 3cac43ed98ff postmaster[528]: 2020-11-25 06:16:08.422 UTC [528] LOG:  PG-Strom: nvme_strom driver is not installed
Nov 25 06:16:08 3cac43ed98ff postmaster[528]: 2020-11-25 06:16:08.422 UTC [528] LOG:  listening on IPv4 address "127.0.0.1", port 5432
Nov 25 06:16:08 3cac43ed98ff postmaster[528]: 2020-11-25 06:16:08.422 UTC [528] LOG:  could not bind IPv6 address "::1": Cannot assign requested address
Nov 25 06:16:08 3cac43ed98ff postmaster[528]: 2020-11-25 06:16:08.422 UTC [528] HINT:  Is another postmaster already running on port 5432? If not, wait a few seconds and retry.
Nov 25 06:16:08 3cac43ed98ff postmaster[528]: 2020-11-25 06:16:08.428 UTC [528] LOG:  listening on Unix socket "/var/run/postgresql/.s.PGSQL.5432"
Nov 25 06:16:08 3cac43ed98ff postmaster[528]: 2020-11-25 06:16:08.439 UTC [528] LOG:  listening on Unix socket "/tmp/.s.PGSQL.5432"
Nov 25 06:16:08 3cac43ed98ff postmaster[528]: 2020-11-25 06:16:08.452 UTC [528] FATAL:  failed on fallocate('/.pg_shmbuf_5432.0:0'): No space left on device
Nov 25 06:16:08 3cac43ed98ff postmaster[528]: 2020-11-25 06:16:08.453 UTC [528] LOG:  database system is shut down
Nov 25 06:16:08 3cac43ed98ff systemd[1]: postgresql-10.service: main process exited, code=exited, status=1/FAILURE
Nov 25 06:16:08 3cac43ed98ff systemd[1]: Failed to start PostgreSQL 10 database server.
-- Subject: Unit postgresql-10.service has failed
-- Defined-By: systemd
-- Support: http://lists.freedesktop.org/mailman/listinfo/systemd-devel
-- 
-- Unit postgresql-10.service has failed.
-- 
-- The result is failed.
Nov 25 06:16:08 3cac43ed98ff systemd[1]: Unit postgresql-10.service entered failed state.
Nov 25 06:16:08 3cac43ed98ff systemd[1]: postgresql-10.service failed.
Nov 25 06:27:22 3cac43ed98ff systemd[1]: Starting PostgreSQL 10 database server...
-- Subject: Unit postgresql-10.service has begun start-up
-- Defined-By: systemd
-- Support: http://lists.freedesktop.org/mailman/listinfo/systemd-devel
-- 
-- Unit postgresql-10.service has begun starting up.
Nov 25 06:27:23 3cac43ed98ff postmaster[574]: 2020-11-25 06:27:23.011 UTC [574] LOG:  NVRTC 11.1 is successfully loaded.
Nov 25 06:27:23 3cac43ed98ff postmaster[574]: 2020-11-25 06:27:23.011 UTC [574] LOG:  PG-Strom version 2.3 built for PostgreSQL 10
Nov 25 06:27:23 3cac43ed98ff postmaster[574]: 2020-11-25 06:27:23.133 UTC [574] LOG:  PG-Strom: GPU0 GeForce GTX 1060 3GB (1152 CUDA cores; 1746MHz, L2 1536kB), RAM 3016MB (192bits, 3.82GHz), CC 6.1
Nov 25 06:27:23 3cac43ed98ff postmaster[574]: 2020-11-25 06:27:23.136 UTC [574] LOG:    -  PCIe[0000:00]
Nov 25 06:27:23 3cac43ed98ff postmaster[574]: 2020-11-25 06:27:23.136 UTC [574] LOG:        -  PCIe(0000:00:03.1)
Nov 25 06:27:23 3cac43ed98ff postmaster[574]: 2020-11-25 06:27:23.136 UTC [574] LOG:            -  PCIe(0000:09:00.0) GPU0 (GeForce GTX 1060 3GB)
Nov 25 06:27:23 3cac43ed98ff postmaster[574]: 2020-11-25 06:27:23.136 UTC [574] LOG:        -  PCIe(0000:00:01.1)
Nov 25 06:27:23 3cac43ed98ff postmaster[574]: 2020-11-25 06:27:23.136 UTC [574] LOG:            -  PCIe(0000:01:00.0) nvme0 (Samsung SSD 980 PRO 500GB)
Nov 25 06:27:23 3cac43ed98ff postmaster[574]: 2020-11-25 06:27:23.136 UTC [574] LOG:  GPU<->SSD Distance Matrix
Nov 25 06:27:23 3cac43ed98ff postmaster[574]: 2020-11-25 06:27:23.136 UTC [574] LOG:             GPU0
Nov 25 06:27:23 3cac43ed98ff postmaster[574]: 2020-11-25 06:27:23.136 UTC [574] LOG:      nvme0     -1
Nov 25 06:27:23 3cac43ed98ff postmaster[574]: 2020-11-25 06:27:23.136 UTC [574] LOG:  Optimal GPUs are uncertain for NVME devices: nvme0
Nov 25 06:27:23 3cac43ed98ff postmaster[574]: 2020-11-25 06:27:23.136 UTC [574] HINT:  review your 'pg_strom.nvme_distance_map' configuration if these devices may be used in PostgreSQL
Nov 25 06:27:23 3cac43ed98ff postmaster[574]: 2020-11-25 06:27:23.138 UTC [574] LOG:  PG-Strom: nvme_strom driver is not installed
Nov 25 06:27:23 3cac43ed98ff postmaster[574]: 2020-11-25 06:27:23.138 UTC [574] LOG:  listening on IPv4 address "127.0.0.1", port 5432
Nov 25 06:27:23 3cac43ed98ff postmaster[574]: 2020-11-25 06:27:23.138 UTC [574] LOG:  could not bind IPv6 address "::1": Cannot assign requested address
Nov 25 06:27:23 3cac43ed98ff postmaster[574]: 2020-11-25 06:27:23.138 UTC [574] HINT:  Is another postmaster already running on port 5432? If not, wait a few seconds and retry.
Nov 25 06:27:23 3cac43ed98ff postmaster[574]: 2020-11-25 06:27:23.144 UTC [574] LOG:  listening on Unix socket "/var/run/postgresql/.s.PGSQL.5432"
Nov 25 06:27:23 3cac43ed98ff postmaster[574]: 2020-11-25 06:27:23.155 UTC [574] LOG:  listening on Unix socket "/tmp/.s.PGSQL.5432"
Nov 25 06:27:23 3cac43ed98ff postmaster[574]: 2020-11-25 06:27:23.169 UTC [574] FATAL:  failed on fallocate('/.pg_shmbuf_5432.0:0'): No space left on device
Nov 25 06:27:23 3cac43ed98ff postmaster[574]: 2020-11-25 06:27:23.170 UTC [574] LOG:  database system is shut down
Nov 25 06:27:23 3cac43ed98ff systemd[1]: postgresql-10.service: main process exited, code=exited, status=1/FAILURE
Nov 25 06:27:23 3cac43ed98ff systemd[1]: Failed to start PostgreSQL 10 database server.
-- Subject: Unit postgresql-10.service has failed
-- Defined-By: systemd
-- Support: http://lists.freedesktop.org/mailman/listinfo/systemd-devel
-- 
-- Unit postgresql-10.service has failed.
-- 
-- The result is failed.
Nov 25 06:27:23 3cac43ed98ff systemd[1]: Unit postgresql-10.service entered failed state.
Nov 25 06:27:23 3cac43ed98ff systemd[1]: postgresql-10.service failed.
lines 1514-1570/1570 (END)
```

**Comment by kaigai** [comment]
Thanks. Could you clean up /dev/shm (including hidden files) before PostgreSQL startup.

PG-Strom, and PostgreSQL also, uses POSIX shared memory that is implemented as a shared-file at /dev/shm.
This error introduces PG-Strom failed to allocate shared memory segment.

**Comment by pmdy** [comment]
/dev/shm looks empty.

```
[root@34d5f5667712 pg-strom-2.3]# cd /dev/shm/
[root@34d5f5667712 shm]# ls
[root@34d5f5667712 shm]# ls
[root@34d5f5667712 shm]# ls -a
.  ..
[root@34d5f5667712 shm]# rm -rf *
```

**Comment by pmdy** [comment]
```
[root@34d5f5667712 dev]# df -h
Filesystem      Size  Used Avail Use% Mounted on
overlay         457G   54G  381G  13% /
tmpfs            64M     0   64M   0% /dev
tmpfs           7.9G     0  7.9G   0% /sys/fs/cgroup
shm              64M     0   64M   0% /dev/shm
/dev/nvme0n1p2  457G   54G  381G  13% /etc/hosts
tmpfs           7.9G   12K  7.9G   1% /proc/driver/nvidia
udev            7.8G     0  7.8G   0% /dev/nvidia0
tmpfs           7.9G  9.1M  7.9G   1% /run
```

**Comment by kaigai** [comment]
> shm              64M     0   64M   0% /dev/shm

64MB!!
Why is it so small? Usually, tmpfs default is half of the system physical memory (and allocates physical page frames on demand).

**Comment by pmdy** [comment]
Haha. Yea, it's docker's default config. Ok, I'll try to resize /dev/shm.

**Comment by pmdy** [comment]
Increase the storage for /dev/shm worked! 

But, I have another roadblock with creating `pg_strom` extension.

```
[root@47ccf998c47d pg-strom-2.3]# psql postgres -U postgres
psql (10.15)
Type "help" for help.

postgres=# CREATE EXTENSION pg_strom;
ERROR:  syntax error at or near "FUNCTION"
postgres=# 
```

**Comment by pmdy** [comment]
Ok, I fixed this issue by downgrading to pg_strom v2.1 from 2.3.
